src/pytorch/pme/pmeCUDA.cu

```diff
@@ -0,0 +1,447 @@
+                          const Scalar& gridx,
+                          const Scalar& gridy,
+                          const Scalar& gridz,
+                          const Scalar& order,
+                          const Scalar& alpha,
+                          const Scalar& coulomb,
+                          const Tensor& xmoduli,
+                          const Tensor& ymoduli,
+                          const Tensor& zmoduli) {
+        const auto stream = c10::cuda::getCurrentCUDAStream(positions.get_device());
+        const c10::cuda::CUDAStreamGuard guard(stream);
+        int numAtoms = positions.size(0);
+        int pmeOrder = (int) order.toInt();
+        int gridSize[3] = {(int) gridx.toInt(), (int) gridy.toInt(), (int) gridz.toInt()};
+        float sqrtCoulomb = sqrt(coulomb.toDouble());
+
+        // Spread the charge on the grid.
+
+        TensorOptions options = torch::TensorOptions().device(positions.device());
+        Tensor realGrid = torch::zeros({gridSize[0], gridSize[1], gridSize[2]}, options);
+        int blockSize = 128;
+        int numBlocks = max(1, min(getMaxBlocks(), (numAtoms+blockSize-1)/blockSize));
+        TORCH_CHECK(pmeOrder == 4 || pmeOrder == 5, "Only pmeOrder 4 or 5 is supported with CUDA");
+        if (pmeOrder == 4)
+            spreadCharge<4><<<numBlocks, blockSize, 0, stream>>>(get_accessor<float, 2>(positions), get_accessor<float, 1>(charges),
+                    get_accessor<float, 2>(box_vectors), get_accessor<float, 3>(realGrid), gridSize[0], gridSize[1], gridSize[2], sqrtCoulomb);
+        else
+            spreadCharge<5><<<numBlocks, blockSize, 0, stream>>>(get_accessor<float, 2>(positions), get_accessor<float, 1>(charges),
+                    get_accessor<float, 2>(box_vectors), get_accessor<float, 3>(realGrid), gridSize[0], gridSize[1], gridSize[2], sqrtCoulomb);
+
+        // Take the Fourier transform.
+
+        Tensor recipGrid = torch::fft::rfftn(realGrid);
+
+        // Perform the convolution and calculate the energy.
+
+        Tensor energy = torch::zeros(numBlocks*blockSize, options);
+        reciprocalConvolution<<<numBlocks, blockSize, 0, stream>>>(get_accessor<float, 2>(box_vectors), get_accessor<c10::complex<float>, 3>(recipGrid),
+                gridSize[0], gridSize[1], gridSize[2], get_accessor<float, 1>(xmoduli), get_accessor<float, 1>(ymoduli), get_accessor<float, 1>(zmoduli),
+                M_PI*M_PI/(alpha.toDouble()*alpha.toDouble()), get_accessor<float, 1>(energy));
+
+        // Store data for later use.
+
+        ctx->save_for_backward({positions, charges, box_vectors, xmoduli, ymoduli, zmoduli, recipGrid});
+        ctx->saved_data["gridx"] = gridx;
+        ctx->saved_data["gridy"] = gridy;
+        ctx->saved_data["gridz"] = gridz;
+        ctx->saved_data["order"] = order;
+        ctx->saved_data["alpha"] = alpha;
+        ctx->saved_data["coulomb"] = coulomb;
+        return {0.5*torch::sum(energy)};
```

Comment:
You would save a kernel launch by putting the 0.5 in reciprocalConvolution, maybe call the tensor half_energy then.

Reply:
Done.